# 01 - Álgebra Lineal con Numpy



En esta parte realizaremos todo lo relacionado con Numpy y sus utilidades en álgebra lineal para saber algunos datos que nos seran de ayuda para realizar el pipeline

Realizare el codigo por secciones para que podamos verificar si se cumple con cada punto pedio en la actividad propuesta 

#### 1. Extracción de datos del csv

In [1]:
# Importamos las librerias que vamos a usar 
import numpy as np  
import pandas as pd

# pandas lee el archivo csv y lo convierte en una tabla
tabla1 = pd.read_csv('../data/ventas_techventas.csv', encoding='latin-1', parse_dates=['fecha'])
tabla1.head() # Primeras 5 filas 

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
0,1001,2024-01-05,C001,TechCorp SA,Norte,Laptop Pro 15,Electronica,2.0,1200.0,0.10,Ana L�ï¿
1,1002,2024-01-07,C002,Innovatek,Sur,Mouse Inalambrico,Perifericos,15.0,25.0,0.00,Carlos Ruiz
2,"1003,2024-01-10,C003,DataSolutions,Centro,Moni...",NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1004,2024-01-12,C001,TechCorp SA,Norte,Teclado Mecanico,Perifericos,10.0,85.0,0.10,Luis Mora
4,1005,2024-01-15,C004,CloudNet,Este,Laptop Pro 15,Electronica,3.0,1200.0,0.15,Carlos Ruiz


#### 1.1 Creacion de arrays

In [2]:
# Seleccionamos las columnas cantidad y precio_unitario por separado y con .to_numpy lasc onvertimos en arrays
cantidades = tabla1["cantidad"].to_numpy()
precios = tabla1["precio_unitario"].to_numpy()

# Prints para verificar las arrays creadas
print("Cantidades:", cantidades)
print("Precios:", precios)

Cantidades: [ 2. 15. nan 10.  3. 20. nan  8. 12.  1. nan 30.  5.  3. 25.  2. nan 18.
 12.  2.  4. 30.  8. nan  4. 20. 15.  6.  1. 10. nan 22.  5. 25.  2. nan
 12. 18. nan  3. 15.  5.  2.  3. 20.  6. 18. -1.  4. 25. nan  2. 20. 35.
 nan  2.  7.  9. 20. nan]
Precios: [1200.   25.   nan   85. 1200.   95.   nan  120.   95. 1200.   nan   25.
   85.  120.   95. 1200.   nan   95.   25. 1200.  120.   95.   85.   nan
 1200.   25.   95.  120. 1200.   85.   nan   95. 1200.   25.  120.   nan
   85.   95.   nan 1200.   25.  120.  350. 1200.   95.   85.   25. 1200.
  120.   95.   nan 1200.   25.   95.   nan 1200.  120.   85.   95.   nan]


### 1.2 Cálculo de ingresos brutos con Numpy (¿Error?)

In [3]:
# Multiplicación elemento a elemento
ingresos_por_producto = cantidades * precios
print("Ingreso por producto:", ingresos_por_producto)

# Producto punto con np.dot (ingreso total)
ingreso_total_dot = np.dot(cantidades, precios)
print("Ingreso total (np.dot):", ingreso_total_dot)

# Lo mismo pero con el operador @
ingreso_total_arroba = cantidades @ precios
print("Ingreso total (@):", ingreso_total_arroba)

# Verificar que ambos dan el mismo resultado
print("¿Son iguales?", ingreso_total_dot == ingreso_total_arroba)

Ingreso por producto: [ 2400.   375.    nan   850.  3600.  1900.    nan   960.  1140.  1200.
    nan   750.   425.   360.  2375.  2400.    nan  1710.   300.  2400.
   480.  2850.   680.    nan  4800.   500.  1425.   720.  1200.   850.
    nan  2090.  6000.   625.   240.    nan  1020.  1710.    nan  3600.
   375.   600.   700.  3600.  1900.   510.   450. -1200.   480.  2375.
    nan  2400.   500.  3325.    nan  2400.   840.   765.  1900.    nan]
Ingreso total (np.dot): nan
Ingreso total (@): nan
¿Son iguales? False


Nos podemos percatar de que hay un error en las operaciones, esto ocurre debido a que, si nos damos cuenta, en las arrays anteiormente creadas podemos ver valores que dicen "nan" lo que es un espacio vacio numericamente, o sea no hay nada escrito en esos espacios del csv donde aparece "nan".

Debido a esto, para completar con este punto de la actividad limpiaremos estos datos con ayuda de Numpy, generalmente se usa la libreria de Pandas pero veo necesario realizar esta pequeña limpieza para continuar.

### 1.2.1 Limpieza de nan

In [4]:
# Eliminar filas donde cantidad o precio sean nan
mask = ~np.isnan(cantidades) & ~np.isnan(precios)
cantidades_limpias = cantidades[mask]
precios_limpios = precios[mask]

# Prints para observar la cantidad de valores que eliminamos con la limpieza de nan
print("Registros originales:", len(cantidades))
print("Registros limpios:", len(cantidades_limpias))

Registros originales: 60
Registros limpios: 49


Con esto ya podremos realizar los cálculos de ingresos brutos de la empresa.

Quiero añadir que dejare el codigo anterior para poder visualizar el error que mencione y volvere a ejecutarlo despues de esta linea de limpieza.

### 1.2 Cálculo de ingresos brutos con Numpy (Corregido)

In [5]:
# Multiplicación elemento a elemento
ingresos_por_producto = cantidades_limpias * precios_limpios
print("Ingreso por producto:", ingresos_por_producto)

# Producto punto con np.dot (ingreso total)
ingreso_total_dot = np.dot(cantidades_limpias, precios_limpios)
print("Ingreso total (np.dot):", ingreso_total_dot)

# Lo mismo pero con el operador @
ingreso_total_arroba = cantidades_limpias @ precios_limpios
print("Ingreso total (@):", ingreso_total_arroba)

# Verificar que ambos dan el mismo resultado
print("¿Son iguales?", ingreso_total_dot == ingreso_total_arroba)

Ingreso por producto: [ 2400.   375.   850.  3600.  1900.   960.  1140.  1200.   750.   425.
   360.  2375.  2400.  1710.   300.  2400.   480.  2850.   680.  4800.
   500.  1425.   720.  1200.   850.  2090.  6000.   625.   240.  1020.
  1710.  3600.   375.   600.   700.  3600.  1900.   510.   450. -1200.
   480.  2375.  2400.   500.  3325.  2400.   840.   765.  1900.]
Ingreso total (np.dot): 73855.0
Ingreso total (@): 73855.0
¿Son iguales? True


Ahora si dan los resultados. 
### Conclusión
Tanto np.dot como el operador @ producen el mismo ingreso bruto total.
La diferencia con * es que este último devuelve un array con el ingreso 
por cada producto, mientras que @ devuelve un único valor total.

### 1.3 Estadísticas con Numpy

In [6]:
# Extraer precio_unitario limpio
precios_stat = tabla1['precio_unitario'].dropna().to_numpy() # .dropna limpia con funcion de la libreria Pandas

# Estadísticas
media = np.mean(precios_stat)
mediana = np.median(precios_stat)
desviacion = np.std(precios_stat)
percentiles = np.percentile(precios_stat, [25, 75])

print("Media:", round(media, 3))                      # round nos ayuda a limitar la cantidad de decimales
print("Mediana:", round(mediana, 3))                   
print("Desviación estándar:", round(desviacion, 3))
print("Percentil 25:", percentiles[0])
print("Percentil 75:", percentiles[1])

Media: 384.592
Mediana: 95.0
Desviación estándar: 492.266
Percentil 25: 85.0
Percentil 75: 1200.0


Podemos observar con los resultados de este estadistico de los datos que el catalogo de TechVentas tiene una distribución de precios asimétrica, con la mayoria de los productos en rangos económicos (al rededor de $85-$95) pero con algunos productos un poco más costosos que elevan significativamente el promedio hasta $384

Quiero quizas dejar un análisis un poco más detallado de los datos optenidos.

### Media $384 vs Mediana $95
La media es mucho más alta que la mediana. Eso significa que hay pocos productos muy caros que jalan la media hacia arriba, pero la mayoría de productos son baratos.

### Desviación estándar $492
Es mayor que la media misma ($384). Eso significa que los precios están muy dispersos, no hay un precio "típico". El catálogo mezcla productos muy baratos con productos muy caros.

### Percentil 25: $85 y Percentil 75: $1200
El 25% de productos cuesta menos de $85 y el 75% cuesta menos de $1200. Esa diferencia tan grande confirma que el catálogo tiene dos rangos muy distintos: productos económicos y productos premium.

## 1.4 Reflexión CRISP-DM

### Fases CRISP-DM

1. Comprensión del negocio:
        TechVentas S.A.S. es una distribuidora de tecnología en Latinoamérica que necesita entender cómo fueron sus ventas en el primer semestre. El CEO requiere respuestas concretas sobre qué productos se vendieron más, qué vendedores tuvieron mejor desempeño y qué regiones generaron más ingresos. Sin este entendimiento previo no tiene sentido analizar ningún dato.

2. Comprensión de los datos:
        Se recibió un dataset con 60 transacciones, 10 clientes, 4 vendedores y 6 meses de datos. Al explorarlo se detectaron problemas como productos en blanco, cantidades negativas y valores nulos. Esta fase es fundamental porque permite saber con qué calidad de datos se trabaja antes de tomar cualquier decisión.

3. Preparación de los datos:
        SEl dataset tiene errores intencionales que deben tratarse antes de cualquier análisis. Se deben eliminar o imputar los valores nulos, descartar las cantidades negativas y verificar que los campos críticos como producto y precio estén completos. Sin esta limpieza los resultados serían incorrectos y engañosos para el CEO.

4. Modelado:
        Con los datos limpios se aplican técnicas de análisis para extraer información útil: cálculo de ingresos brutos y netos, estadísticas descriptivas de precios, consultas SQL para agrupar por vendedor y región, y comparaciones de cumplimiento de metas.

5. Evaluación:
        Se verifica que los resultados respondan las preguntas originales del negocio. ¿Los ingresos calculados tienen sentido? ¿Los vendedores con mejor desempeño coinciden con lo esperado? ¿Las regiones más rentables son coherentes con la operación de la empresa?

6. Despliegue:
        El proyecto se entrega como un repositorio público en GitHub con notebooks documentados, código comentado y un README claro para que cualquier persona pueda reproducir el análisis desde cero.

## ¿Cuál fase es la más importante?

Si bien todas las fases son necesarias y dependen una de la otra, considero que la fase más importante en este reto es la Preparación de los datos.
En un proyecto real de datos, se estima que un analista dedica entre el 60% y 80% de su tiempo a limpiar y preparar datos, y este reto lo refleja perfectamente. El dataset de TechVentas tiene errores intencionales: productos en blanco, cantidades negativas y valores nulos. Si estos errores no se detectan y tratan correctamente, todos los análisis posteriores estarían construidos sobre una base incorrecta.
Sin embargo, también podría argumentarse que la fase de Comprensión del negocio es igual de importante, porque si no se entiende qué preguntas necesita responder el CEO, se podrían limpiar y analizar los datos correctamente pero responder las preguntas equivocadas. En el mundo real ambas fases van de la mano.